elastic/kibana
# [Lens] Fix cross cluster bug and optimize existence query

This fixes a bug which came up during my testing of Lens in cloud. Cloud makes it easy to set up a cross-cluster-search node, but that caused the following bug:

1. Create a cluster in cloud
2. Create a second cluster with cross-cluster-search
3. On the CCS cluster, create an index pattern such as first-cluster:kibana_sample_data_logs
4. Open this in Lens and you'll see a totally empty list of fields

This PR fixes the bug by detecting CCS index patterns which use the `:` character, as seen in this screenshot:

<img width="972" alt="Screenshot 2020-06-26 17 51 34" src="https://user-images.githubusercontent.com/666475/85904126-fc7a1e00-b7d5-11ea-9cad-ad6cb0cf6330.png">

Because it touched the same code, I've also optimized the search query we use to determine field existence. We need to sort by the timestamp field to take advantage of Elasticsearch's ability to skip older shards.

Fixes https://github.com/elastic/kibana/issues/69460
Fixes https://github.com/elastic/kibana/issues/70017

## Local reproduction steps

This PR was hard to test locally. Here are the steps I used:

1. Download a build of Elasticsearch and run it with the default settings. This will be the "remote cluster"
2. Once the build is running, create an index and sample document with these commands:

```
curl -u elastic -p -X PUT localhost:9200/sample_data -H 'Content-Type: application/json' -d'{
  "mappings": {
    "properties" : {
      "test" : {
        "type" : "keyword",
        "ignore_above" : 256
      }
    }
  }
}'
```

```
curl -u elastic -p -X POST localhost:9200/sample_data/_doc -H 'Content-Type: application/json' -d'{
  "test": "Hello World!"
}'
```

3. Start a new ES snapshot using `yarn es snapshot -E http.port=9501 -E transport.port=9601`
4. Start Kibana with `yarn start --dev --elasticsearch.hosts=http://localhost:9501`
5. Open the Kibana dev tools, and run this command:

```
PUT _cluster/settings
{
  "persistent": {
    "cluster": {
      "remote": {
        "cluster_one": {
          "seeds": [
            "127.0.0.1:9300"
          ],
          "transport.ping_schedule": "30s"
        }
      }
    }
  }
}
```

6. Create a new index pattern object with the name `cluster_one:sample_data`
7. Check for the field `test` to appear in your index pattern. If not, you will need to follow these steps (this works automatically on cloud):
   a. Using the Kibana dev tools, find the CCS index pattern with this query:

```
GET .kibana/_search
{
  "query": {
    "term": { "index-pattern.title": "cluster_one:sample_data" }
  }
}
```

    b. Copy this command in, changing the ID to the `_id` of the document you created.

```
PUT .kibana/_doc/<ID here>
{
          "index-pattern" : {
            "title" : "cluster_one:sample_data",
            "fields" : """[{"name":"_id","type":"string","count":0,"scripted":false,"searchable":false,"aggregatable":false,"readFromDocValues":false},{"name":"_index","type":"string","count":0,"scripted":false,"searchable":false,"aggregatable":false,"readFromDocValues":false},{"name":"_score","type":"number","count":0,"scripted":false,"searchable":false,"aggregatable":false,"readFromDocValues":false},{"name":"_source","type":"_source","count":0,"scripted":false,"searchable":false,"aggregatable":false,"readFromDocValues":false},{"name":"_type","type":"string","count":0,"scripted":false,"searchable":false,"aggregatable":false,"readFromDocValues":false},{"name":"test","type":"string","count":0,"scripted":false,"searchable":true,"aggregatable":true,"readFromDocValues":true}]"""
          },
          "type" : "index-pattern",
          "references" : [ ],
          "migrationVersion" : {
            "index-pattern" : "7.6.0"
          },
          "updated_at" : "2020-06-26T21:24:25.559Z"
      }
```

Finally, test in Lens. You should see the field "test" is available.
